scipy/optimize/_lsq/givens_elimination.pyx

```diff
@@ -0,0 +1,70 @@
+from libc.math cimport sqrt, fabs, copysign
+cimport numpy as np
+
+import numpy as np
+
+
+def givens_elimination(double [:, :] S, double [:] v, double [:] diag):
+    """Apply a series of Givens rotations to zero out a diagonal block of
+    a matrix.
+
+    The matrix has the structure::
+
+        [ S ]
+        [ D ]
+
+    Where S is an upper triangular matrix with shape (n, n) and D is a
+    diagonal matrix with shape (n, n) with elements from `diag`. This function
+    applies Givens rotations to it such that the resulting matrix has zeros
+    in place of D.
+
+    Array `S` will be modified in-place.
+
+    Array `v` of shape (n,) is the part of the full vector with shape (2*n,)::
+
+        [ v ]
+        [ 0 ]
+
+    to which Givens rotations are applied. This array is modified in place,
+    such that on exit it contains the first n components of the above
+    mentioned vector after rotations were applied.
+    """
+    cdef int n = diag.size
```

Comment:
Yes, this looks fine.  All arguments are references because both BLAS and LAPACK are natively  Fortran APIs and Fortran is always pass-by-reference.